File: tools/regression/bart/individual_regression_test_bart.py
import numpy as np
import pandas as pd
import time
import sys
import os
from typing import Dict
from stochtree import BARTModel
from sklearn.model_selection import train_test_split

def dgp1(n: int, p: int, snr: float) -> Dict:
    rng = np.random.default_rng()
    
    # Covariates
    X = rng.uniform(0, 1, size=(n, p))
    
    # Piecewise linear term
    plm_term = np.where(
        (X[:,0] >= 0.0) & (X[:,0] < 0.25), -7.5 * X[:,1], 
        np.where(
            (X[:,0] >= 0.25) & (X[:,0] < 0.5), -2.5 * X[:,1], 
            np.where(
                (X[:,0] >= 0.5) & (X[:,0] < 0.75), 2.5 * X[:,1], 
                7.5 * X[:,1]
            )
        )
    )
    
    # Trigonometric term
    trig_term = 2 * np.sin(X[:, 2] * 2 * np.pi) - 1.5 * np.cos(X[:, 3] * 2 * np.pi)
    
    # Outcome
    f_XW = plm_term + trig_term
    noise_sd = np.std(f_XW) / snr
    y = f_XW + rng.normal(0, noise_sd, n)
    
    return {
        'covariates': X,
        'basis': None,
        'outcome': y,
        'conditional_mean': f_XW,
        'rfx_group_ids': None,
        'rfx_basis': None
    }


def dgp2(n: int, p: int, snr: float) -> Dict:
    rng = np.random.default_rng()
    
    # Covariates and basis
    X = rng.uniform(0, 1, (n, p))
    W = rng.uniform(0, 1, (n, 2))
    
    # Piecewise linear term using basis W
    plm_term = np.where(
        (X[:,0] >= 0.0) & (X[:,0] < 0.25), -7.5 * W[:,0], 
        np.where(
            (X[:,0] >= 0.25) & (X[:,0] < 0.5), -2.5 * W[:,0], 
            np.where(
                (X[:,0] >= 0.5) & (X[:,0] < 0.75), 2.5 * W[:,0], 
                7.5 * W[:,0]
            )
        )
    )
    
    # Trigonometric term
    trig_term = 2 * np.sin(X[:, 2] * 2 * np.pi) - 1.5 * np.cos(X[:, 3] * 2 * np.pi)
    
    # Outcome
    f_XW = plm_term + trig_term
    noise_sd = np.std(f_XW) / snr
    y = f_XW + rng.normal(0, noise_sd, n)
    
    return {
        'covariates': X,
        'basis': W,
        'outcome': y,
        'conditional_mean': f_XW,
        'rfx_group_ids': None,
        'rfx_basis': None
    }


def dgp3(n: int, p: int, snr: float) -> Dict:
    rng = np.random.default_rng()
    
    # Covariates
    X = rng.uniform(0, 1, size=(n, p))
    
    # Piecewise linear term
    plm_term = np.where(
        (X[:,0] >= 0.0) & (X[:,0] < 0.25), -7.5 * X[:,1], 
        np.where(
            (X[:,0] >= 0.25) & (X[:,0] < 0.5), -2.5 * X[:,1], 
            np.where(
                (X[:,0] >= 0.5) & (X[:,0] < 0.75), 2.5 * X[:,1], 
                7.5 * X[:,1]
            )
        )
    )
    
    # Trigonometric term
    trig_term = 2 * np.sin(X[:, 2] * 2 * np.pi) - 1.5 * np.cos(X[:, 3] * 2 * np.pi)
    
    # Random effects
    num_groups = 3
    rfx_group_ids = rng.choice(num_groups, size=n)
    rfx_coefs = np.array([[-5, -3, -1], [5, 3, 1]]).T
    rfx_basis = np.column_stack([np.ones(n), np.random.uniform(-1, 1, n)])
    rfx_term = np.sum(rfx_coefs[rfx_group_ids] * rfx_basis, axis=1)
    
    # Outcome
    f_XW = plm_term + trig_term + rfx_term
    noise_sd = np.std(f_XW) / snr
    y = f_XW + rng.normal(0, noise_sd, n)
    
    return {
        'covariates': X,
        'basis': None,
        'outcome': y,
        'conditional_mean': f_XW,
        'rfx_group_ids': rfx_group_ids,
        'rfx_basis': rfx_basis
    }


def dgp4(n: int, p: int, snr: float) -> Dict:
    rng = np.random.default_rng()
    
    # Covariates and basis
    X = rng.uniform(0, 1, (n, p))
    W = rng.uniform(0, 1, (n, 2))
    
    # Piecewise linear term using basis W
    plm_term = np.where(
        (X[:,0] >= 0.0) & (X[:,0] < 0.25), -7.5 * W[:,0], 
        np.where(
            (X[:,0] >= 0.25) & (X[:,0] < 0.5), -2.5 * W[:,0], 
            np.where(
                (X[:,0] >= 0.5) & (X[:,0] < 0.75), 2.5 * W[:,0], 
                7.5 * W[:,0]
            )
        )
    )
    
    # Trigonometric term
    trig_term = 2 * np.sin(X[:, 2] * 2 * np.pi) - 1.5 * np.cos(X[:, 3] * 2 * np.pi)
    
    # Random effects
    num_groups = 3
    rfx_group_ids = rng.choice(num_groups, size=n)
    rfx_coefs = np.array([[-5, -3, -1], [5, 3, 1]]).T
    rfx_basis = np.column_stack([np.ones(n), np.random.uniform(-1, 1, n)])
    rfx_term = np.sum(rfx_coefs[rfx_group_ids] * rfx_basis, axis=1)
    
    # Outcome
    f_XW = plm_term + trig_term + rfx_term
    noise_sd = np.std(f_XW) / snr
    y = f_XW + np.random.normal(0, noise_sd, n)
    
    return {
        'covariates': X,
        'basis': W,
        'outcome': y,
        'conditional_mean': f_XW,
        'rfx_group_ids': rfx_group_ids,
        'rfx_basis': rfx_basis
    }


def compute_test_train_indices(n: int, test_set_pct: float) -> Dict[str, np.ndarray]:
    sample_inds = np.arange(n)
    train_inds, test_inds = train_test_split(sample_inds, test_size=test_set_pct)
    return {'test_inds': test_inds, 'train_inds': train_inds}


def subset_data(data: np.ndarray, subset_inds: np.ndarray) -> np.ndarray:
    if isinstance(data, np.ndarray):
        if data.ndim == 1:
            return data[subset_inds]
        else:
            return data[subset_inds, :]
    else:
        raise ValueError("Data must be a numpy array")


def main():
    # Parse command line arguments
    if len(sys.argv) > 1:
        n_iter = int(sys.argv[1])
        n = int(sys.argv[2])
        p = int(sys.argv[3])
        num_gfr = int(sys.argv[4])
        num_mcmc = int(sys.argv[5])
        dgp_num = int(sys.argv[6])
        snr = float(sys.argv[7])
        test_set_pct = float(sys.argv[8])
        num_threads = int(sys.argv[9])
    else:
        # Default arguments
        n_iter = 5
        n = 1000
        p = 5
        num_gfr = 10
        num_mcmc = 100
        dgp_num = 1
        snr = 2.0
        test_set_pct = 0.2
        num_threads = -1
    
    print(f"n_iter = {n_iter}")
    print(f"n = {n}")
    print(f"p = {p}")
    print(f"num_gfr = {num_gfr}")
    print(f"num_mcmc = {num_mcmc}")
    print(f"dgp_num = {dgp_num}")
    print(f"snr = {snr}")
    print(f"test_set_pct = {test_set_pct}")
    print(f"num_threads = {num_threads}")
    
    # Run the performance evaluation
    results = np.empty((n_iter, 4), dtype=float)
    
    for i in range(n_iter):
        print(f"Running iteration {i+1}/{n_iter}")
        
        # Generate data
        if dgp_num == 1:
            data_dict = dgp1(n=n, p=p, snr=snr)
        elif dgp_num == 2:
            data_dict = dgp2(n=n, p=p, snr=snr)
        elif dgp_num == 3:
            data_dict = dgp3(n=n, p=p, snr=snr)
        elif dgp_num == 4:
            data_dict = dgp4(n=n, p=p, snr=snr)
        else:
            raise ValueError("Invalid DGP input")
        
        covariates = data_dict['covariates']
        basis = data_dict['basis']
        conditional_mean = data_dict['conditional_mean']
        outcome = data_dict['outcome']
        rfx_group_ids = data_dict['rfx_group_ids']
        rfx_basis = data_dict['rfx_basis']
        
        # Split into train / test sets
        subset_inds_dict = compute_test_train_indices(n, test_set_pct)
        test_inds = subset_inds_dict['test_inds']
        train_inds = subset_inds_dict['train_inds']
        covariates_train = subset_data(covariates, train_inds)
        covariates_test = subset_data(covariates, test_inds)
        outcome_train = subset_data(outcome, train_inds)
        outcome_test = subset_data(outcome, test_inds)
        conditional_mean_train = subset_data(conditional_mean, train_inds)
        conditional_mean_test = subset_data(conditional_mean, test_inds)
        has_basis = basis is not None
        has_rfx = rfx_group_ids is not None
        if has_basis:
            basis_train = subset_data(basis, train_inds)
            basis_test = subset_data(basis, test_inds)
        else:
            basis_train = None
            basis_test = None
        if has_rfx:
            rfx_group_ids_train = subset_data(rfx_group_ids, train_inds)
            rfx_group_ids_test = subset_data(rfx_group_ids, test_inds)
            rfx_basis_train = subset_data(rfx_basis, train_inds)
            rfx_basis_test = subset_data(rfx_basis, test_inds)
        else:
            rfx_group_ids_train = None
            rfx_group_ids_test = None
            rfx_basis_train = None
            rfx_basis_test = None
        
        # Run (and time) BART
        start_time = time.time()
        
        # Sample BART model
        general_params = {'num_threads': num_threads}
        bart_model = BARTModel()
        bart_model.sample(
            X_train=covariates_train,
            y_train=outcome_train,
            leaf_basis_train=basis_train,
            rfx_group_ids_train=rfx_group_ids_train,
            rfx_basis_train=rfx_basis_train,
            num_gfr=num_gfr,
            num_mcmc=num_mcmc,
            general_params=general_params
        )
        
        # Predict on the test set
        test_preds = bart_model.predict(
            X=covariates_test,
            leaf_basis=basis_test,
            rfx_group_ids=rfx_group_ids_test,
            rfx_basis=rfx_basis_test
        )
        
        bart_timing = time.time() - start_time
        
        # Compute test set evaluations
        y_hat_posterior = test_preds['y_hat']
        y_hat_posterior_mean = np.mean(y_hat_posterior, axis=1)
        rmse_test = np.sqrt(np.mean((y_hat_posterior_mean - conditional_mean_test) ** 2))
        
        y_hat_posterior_quantile_025 = np.percentile(y_hat_posterior, 2.5, axis=1)
        y_hat_posterior_quantile_975 = np.percentile(y_hat_posterior, 97.5, axis=1)
        
        covered = np.logical_and(
            conditional_mean_test >= y_hat_posterior_quantile_025,
            conditional_mean_test <= y_hat_posterior_quantile_975
        )
        coverage_test = np.mean(covered)
        
        # Store evaluations
        results[i, :] = [i+1, rmse_test, coverage_test, bart_timing]
    
    # Wrangle and save results to CSV
    results_df = pd.DataFrame({
        'n': n,
        'p': p,
        'num_gfr': num_gfr,
        'num_mcmc': num_mcmc,
        'dgp_num': dgp_num,
        'snr': snr,
        'test_set_pct': test_set_pct,
        'num_threads': num_threads,
        'iter': results[:, 0],
        'rmse': results[:, 1],
        'coverage': results[:, 2],
        'runtime': results[:, 3]
    })
    
    snr_rounded = int(snr)
    test_set_pct_rounded = int(test_set_pct * 100)
    num_threads_clean = 0 if num_threads < 0 else num_threads
    filename = f"stochtree_bart_python_n_{n}_p_{p}_num_gfr_{num_gfr}_num_mcmc_{num_mcmc}_dgp_num_{dgp_num}_snr_{snr_rounded}_test_set_pct_{test_set_pct_rounded}_num_threads_{num_threads_clean}.csv"
    output_dir = "tools/regression/bart/stochtree_bart_python_results"
    filename_full = os.path.join(output_dir, filename)
    results_df.to_csv(filename_full, index=False)


if __name__ == "__main__":
    main() 

File: tools/regression/bart/individual_regression_test_bart.R
# Load libraries
library(stochtree)

# Define DGPs
dgp1 <- function(n, p, snr) {
    X <- matrix(runif(n*p), ncol = p)
    plm_term <- (
        ((0 <= X[,1]) & (0.25 > X[,1])) * (-7.5*X[,2]) + 
        ((0.25 <= X[,1]) & (0.5 > X[,1])) * (-2.5*X[,2]) + 
        ((0.5 <= X[,1]) & (0.75 > X[,1])) * (2.5*X[,2]) + 
        ((0.75 <= X[,1]) & (1 > X[,1])) * (7.5*X[,2])
    )
    trig_term <- (
        2*sin(X[,3]*2*pi) - 
        1.5*cos(X[,4]*2*pi)
    )
    f_XW <- plm_term + trig_term
    noise_sd <- sd(f_XW)/snr
    y <- f_XW + rnorm(n, 0, noise_sd)
    return(list(covariates = X, basis = NULL, outcome = y, conditional_mean = f_XW, 
                rfx_group_ids = NULL, rfx_basis = NULL))
}
dgp2 <- function(n, p, snr) {
    X <- matrix(runif(n*p), ncol = p)
    W <- matrix(runif(n*2), ncol = 2)
    plm_term <- (
        ((0 <= X[,1]) & (0.25 > X[,1])) * (-7.5*W[,1]) + 
        ((0.25 <= X[,1]) & (0.5 > X[,1])) * (-2.5*W[,1]) + 
        ((0.5 <= X[,1]) & (0.75 > X[,1])) * (2.5*W[,1]) + 
        ((0.75 <= X[,1]) & (1 > X[,1])) * (7.5*W[,1])
    )
    trig_term <- (
        2*sin(X[,3]*2*pi) - 
        1.5*cos(X[,4]*2*pi)
    )
    f_XW <- plm_term + trig_term
    noise_sd <- sd(f_XW)/snr
    y <- f_XW + rnorm(n, 0, noise_sd)
    return(list(covariates = X, basis = W, outcome = y, conditional_mean = f_XW, 
                rfx_group_ids = NULL, rfx_basis = NULL))
}
dgp3 <- function(n, p, snr) {
    X <- matrix(runif(n*p), ncol = p)
    plm_term <- (
        ((0 <= X[,1]) & (0.25 > X[,1])) * (-7.5*X[,2]) + 
        ((0.25 <= X[,1]) & (0.5 > X[,1])) * (-2.5*X[,2]) + 
        ((0.5 <= X[,1]) & (0.75 > X[,1])) * (2.5*X[,2]) + 
        ((0.75 <= X[,1]) & (1 > X[,1])) * (7.5*X[,2])
    )
    trig_term <- (
        2*sin(X[,3]*2*pi) - 
        1.5*cos(X[,4]*2*pi)
    )
    rfx_group_ids <- sample(1:3, size = n, replace = T)
    rfx_coefs <- t(matrix(c(-5, -3, -1, 5, 3, 1), nrow=2, byrow=TRUE))
    rfx_basis <- cbind(1, runif(n, -1, 1))
    rfx_term <- rowSums(rfx_coefs[rfx_group_ids,] * rfx_basis)
    f_XW <- plm_term + trig_term + rfx_term
    noise_sd <- sd(f_XW)/snr
    y <- f_XW + rnorm(n, 0, noise_sd)
    return(list(covariates = X, basis = NULL, outcome = y, conditional_mean = f_XW, 
                rfx_group_ids = rfx_group_ids, rfx_basis = rfx_basis))
}
dgp4 <- function(n, p, snr) {
    X <- matrix(runif(n*p), ncol = p)
    W <- matrix(runif(n*2), ncol = 2)
    plm_term <- (
        ((0 <= X[,1]) & (0.25 > X[,1])) * (-7.5*W[,1]) + 
        ((0.25 <= X[,1]) & (0.5 > X[,1])) * (-2.5*W[,1]) + 
        ((0.5 <= X[,1]) & (0.75 > X[,1])) * (2.5*W[,1]) + 
        ((0.75 <= X[,1]) & (1 > X[,1])) * (7.5*W[,1])
    )
    trig_term <- (
        2*sin(X[,3]*2*pi) - 
        1.5*cos(X[,4]*2*pi)
    )
    rfx_group_ids <- sample(1:3, size = n, replace = T)
    rfx_coefs <- t(matrix(c(-5, -3, -1, 5, 3, 1), nrow=2, byrow=TRUE))
    rfx_basis <- cbind(1, runif(n, -1, 1))
    rfx_term <- rowSums(rfx_coefs[rfx_group_ids,] * rfx_basis)
    f_XW <- plm_term + trig_term + rfx_term
    noise_sd <- sd(f_XW)/snr
    y <- f_XW + rnorm(n, 0, noise_sd)
    return(list(covariates = X, basis = W, outcome = y, conditional_mean = f_XW, 
                rfx_group_ids = rfx_group_ids, rfx_basis = rfx_basis))
}

# Test / train split utilities
compute_test_train_indices <- function(n, test_set_pct) {
    n_test <- round(test_set_pct*n)
    n_train <- n - n_test
    test_inds <- sort(sample(1:n, n_test, replace = FALSE))
    train_inds <- (1:n)[!((1:n) %in% test_inds)]
    return(list(test_inds = test_inds, train_inds = train_inds))
}
subset_data <- function(data, subset_inds) {
    if (is.matrix(data)) {
        return(data[subset_inds,])
    } else {
        return(data[subset_inds])
    }
}

# Capture command line arguments
args <- commandArgs(trailingOnly = T)
if (length(args) > 0){
    n_iter <- as.integer(args[1])
    n <- as.integer(args[2])
    p <- as.integer(args[3])
    num_gfr <- as.integer(args[4])
    num_mcmc <- as.integer(args[5])
    dgp_num <- as.integer(args[6])
    snr <- as.numeric(args[7])
    test_set_pct <- as.numeric(args[8])
    num_threads <- as.integer(args[9])
} else{
    # Default arguments
    n_iter <- 5
    n <- 1000
    p <- 5
    num_gfr <- 10
    num_mcmc <- 100
    dgp_num <- 1
    snr <- 2.0
    test_set_pct <- 0.2
    num_threads <- -1
}
cat("n_iter = ", n_iter, "\nn = ", n, "\np = ", p, "\nnum_gfr = ", num_gfr, 
    "\nnum_mcmc = ", num_mcmc, "\ndgp_num = ", dgp_num, "\nsnr = ", snr, 
    "\ntest_set_pct = ", test_set_pct, "\nnum_threads = ", num_threads, "\n", sep = "")

# Run the performance evaluation
results <- matrix(NA, nrow = n_iter, ncol = 4)
colnames(results) <- c("iter", "rmse", "coverage", "runtime")
for (i in 1:n_iter) {
    # Generate data
    if (dgp_num == 1) {
        data_list <- dgp1(n = n, p = p, snr = snr)
    } else if (dgp_num == 2) {
        data_list <- dgp2(n = n, p = p, snr = snr)
    } else if (dgp_num == 3) {
        data_list <- dgp3(n = n, p = p, snr = snr)
    } else if (dgp_num == 4) {
        data_list <- dgp4(n = n, p = p, snr = snr)
    } else {
        stop("Invalid DGP input")
    }
    covariates <- data_list[['covariates']]
    basis <- data_list[['basis']]
    conditional_mean <- data_list[['conditional_mean']]
    outcome <- data_list[['outcome']]
    rfx_group_ids <- data_list[['rfx_group_ids']]
    rfx_basis <- data_list[['rfx_basis']]
    
    # Split into train / test sets
    subset_inds_list <- compute_test_train_indices(n, test_set_pct)
    test_inds <- subset_inds_list$test_inds
    train_inds <- subset_inds_list$train_inds
    covariates_train <- subset_data(covariates, train_inds)
    covariates_test <- subset_data(covariates, test_inds)
    outcome_train <- subset_data(outcome, train_inds)
    outcome_test <- subset_data(outcome, test_inds)
    conditional_mean_train <- subset_data(conditional_mean, train_inds)
    conditional_mean_test <- subset_data(conditional_mean, test_inds)
    has_basis <- !is.null(basis)
    has_rfx <- !is.null(rfx_group_ids)
    if (has_basis) {
        basis_train <- subset_data(basis, train_inds)
        basis_test <- subset_data(basis, test_inds)
    } else {
        basis_train <- NULL
        basis_test <- NULL
    }
    if (has_rfx) {
        rfx_group_ids_train <- subset_data(rfx_group_ids, train_inds)
        rfx_group_ids_test <- subset_data(rfx_group_ids, test_inds)
        rfx_basis_train <- subset_data(rfx_basis, train_inds)
        rfx_basis_test <- subset_data(rfx_basis, test_inds)
    } else {
        rfx_group_ids_train <- NULL
        rfx_group_ids_test <- NULL
        rfx_basis_train <- NULL
        rfx_basis_test <- NULL
    }
    
    # Run (and time) BART
    bart_timing <- system.time({
        # Sample BART model
        general_params <- list(num_threads = num_threads)
        bart_model <- stochtree::bart(
            X_train = covariates_train, y_train = outcome_train, leaf_basis_train = basis_train, 
            rfx_group_ids_train = rfx_group_ids_train, rfx_basis_train = rfx_basis_train, 
            num_gfr = num_gfr, num_mcmc = num_mcmc, general_params = general_params
        )
        
        # Predict on the test set
        test_preds <- predict(
            bart_model, X = covariates_test, leaf_basis = basis_test, 
            rfx_group_ids = rfx_group_ids_test, rfx_basis = rfx_basis_test
        )
    })[3]
    
    # Compute test set evals
    y_hat_posterior <- test_preds$y_hat
    y_hat_posterior_mean <- rowMeans(y_hat_posterior)
    rmse_test <- sqrt(mean((y_hat_posterior_mean - outcome_test)^2))
    y_hat_posterior_quantile_025 <- apply(y_hat_posterior, 1, function(x) quantile(x, 0.025))
    y_hat_posterior_quantile_975 <- apply(y_hat_posterior, 1, function(x) quantile(x, 0.975))
    covered <- rep(NA, nrow(y_hat_posterior))
    for (j in 1:nrow(y_hat_posterior)) {
        covered[j] <- (
            (conditional_mean_test[j] >= y_hat_posterior_quantile_025[j]) & 
            (conditional_mean_test[j] <= y_hat_posterior_quantile_975[j])
        )
    }
    coverage_test <- mean(covered)

    # Store evaluations
    results[i,] <- c(i, rmse_test, coverage_test, bart_timing)
}

# Wrangle and save results to CSV
results_df <- data.frame(
    cbind(n, p, num_gfr, num_mcmc, dgp_num, snr, test_set_pct, num_threads, results)
)
snr_rounded <- as.integer(snr)
test_set_pct_rounded <- as.integer(test_set_pct*100)
num_threads_clean <- ifelse(num_threads < 0, 0, num_threads)
filename <- paste(
    "stochtree", "bart", "r", "n", n, "p", p, "num_gfr", num_gfr, "num_mcmc", num_mcmc, 
    "dgp_num", dgp_num, "snr", snr_rounded, "test_set_pct", test_set_pct_rounded, 
    "num_threads", num_threads_clean, sep = "_"
)
filename_full <- paste0("tools/regression/bart/stochtree_bart_r_results/", filename, ".csv")
write.csv(x = results_df, file = filename_full, row.names = F)
